### PR TITLE
LibWeb: Improve find in page performance on large documents

### PR DESCRIPTION
**Pull request by tcl3**
This PR makes 2 changes, which greatly improve the performance of find in page on pages with a lot of text. 

See the descriptions of the individual commits for details.

There is still more work that could be done to improve performance further. Case-insensitive matching is still slow on documents containing very large text blocks, such as [this copy](https://raw.githubusercontent.com/mmcky/nyu-econ-370/master/notebooks/data/book-war-and-peace.txt) of War & Peace.

I've included a test on one of the longest articles on Wikipedia.

Before:

https://github.com/LadybirdBrowser/ladybird/assets/2817754/b825e5ac-9d25-4225-a1d3-c0bbbd2dd981

After:

https://github.com/LadybirdBrowser/ladybird/assets/2817754/d0bd32ae-747f-4999-ad8e-103509755cac

Note: For both of the above tests, I waited a good 30 seconds for the page to finish loading before doing the query. Find in page still performs poorly if done while the document is partially loaded.

### CODE REVIEW

File: Userland/Libraries/LibWeb/Layout/Viewport.cpp [hunk 1]
```diff
@@ -27,4 +27,55 @@ JS::GCPtr<Painting::Paintable> Viewport::create_paintable() const
...
+{
+    StringBuilder builder;
+    size_t current_start_position = 0;
+    Vector<TextPosition> text_positions;
+    Vector<TextBlock> text_blocks;
+    for_each_in_inclusive_subtree([&](auto const& layout_node) {
+        if (layout_node.display().is_none() || !layout_node.paintable() || !layout_node.paintable()->is_visible())
+            return TraversalDecision::Continue;
+
+        if (layout_node.is_block_container()) {
```
Comment (kalenikaliaksandr) [inline Userland/Libraries/LibWeb/Layout/Viewport.cpp:48]: hmm, if block container interrupts TextBlock then shouldn't the same happen for grid, flex and table containers?

File: Userland/Libraries/LibWeb/DOM/Document.cpp [hunk 2]
```diff
@@ -5111,65 +5111,13 @@ void Document::set_needs_to_refresh_scroll_state(bool b)
...
-            text_blocks.append({ builder.to_string_without_validation(), text_positions });
-
-        return text_blocks;
-    };
-
     // Ensure the layout tree exists before searching for text matches.
     update_layout();
 
-    auto text_blocks = gather_text_blocks();
+    auto text_blocks = layout_node()->text_blocks();
```
Comment (kalenikaliaksandr) [inline Userland/Libraries/LibWeb/DOM/Document.cpp:5120]: `text_blocks` vector is copied here. let's make `text_blocks()` return `const&` instead.

File: Userland/Libraries/LibWeb/Layout/Viewport.h [hunk 3]
```diff
@@ -19,12 +19,26 @@ class Viewport final : public BlockContainer {
     explicit Viewport(DOM::Document&, NonnullRefPtr<CSS::StyleProperties>);
     virtual ~Viewport() override;
 
+    struct TextPosition {
+        DOM::Text& dom_node;
```
Comment (kalenikaliaksandr) [inline Userland/Libraries/LibWeb/Layout/Viewport.h:23]: I know, it's probably safe to assume that DOM nodes will live for at least as long as layout tree, but let's always visit GC-allocated members :)

File: Userland/Libraries/LibWeb/Layout/Viewport.cpp (hunk 1 above)
Comment (tcl3) [reply Userland/Libraries/LibWeb/Layout/Viewport.cpp]: I've changed the check here to be `is_box()` instead. I *think* this covers all scenarios where we'd want a split to occur.

File: Userland/Libraries/LibWeb/DOM/Document.cpp (hunk 2 above)
Comment (kalenikaliaksandr) [inline Userland/Libraries/LibWeb/DOM/Document.cpp:5120]:
```suggestion
    auto const& text_blocks = layout_node()->text_blocks();
```

Comment (tcl3) [reply Userland/Libraries/LibWeb/DOM/Document.cpp:5120]: I realized I didn't do that immediately after I pushed :facepalm:

**Pull request merged**